File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        
        for(int i = nums.length - 1, q = 1; i >= 0; i--){
            res[i] = q;
            q = q * nums[i];
        }

        for(int i = 0, p = 1; i < res.length; i++){
            res[i] = p * res[i];
            p = p * nums[i];
        }

        return res;
    }
}  

File: Data Structures & Algorithms/string-encode-and-decode/submission-1.py
class Solution:
    def encode(self, strs: List[str]) -> str:
        res = ""
        for s in strs:
            res += str(len(s)) + "#" + s

        return res

    # 1#a 2#ab
    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        while i < len(s):
            length = 0

            while s[i] != "#":
                length = length * 10 + int(s[i])
                i += 1

            i += 1
            res.append(s[i : i + length])
            i += length

        return res
